### PR TITLE
Implement values for hubble-relay to properly control sub chart values

### LINKED ISSUE #11755: Helm: Fix values scoping for newly implemented sub-chart
## Bug report

Helm sub-chart for hubble-relay was implemented recently using values that don't control the sub-charts properly. Fix the sub-chart for hubble-relay so values are implemented at the top-level and can properly control the sub-charts. 

Implement defaults in the sub-charts values files. Document overrides in the parent chart values.

https://docs.helm.sh/docs/chart_template_guide/subcharts_and_globals/#overriding-values-from-a-parent-chart

### PR DESCRIPTION
**Pull request by seanmwinn**
Implements hubble-relay values to properly control sub chart values using the parent chart's values file.
Documents values in the parent chart for overriding.

Fixes: #11755

Signed-off-by: Sean Winn <sean@isovalent.com>

### CODE REVIEW

File: install/kubernetes/cilium/charts/hubble-cli/templates/daemonset.yaml
Comment (gandro) [inline install/kubernetes/cilium/charts/hubble-cli/templates/daemonset.yaml]: This changes the test condition (hubble.enabled vs hubble-cli enabled). What's the rationale for this change?

I personally think we can also remove this if. In hubble-relay and ui we have the validation.yaml instead.

File: install/kubernetes/cilium/charts/hubble-ui/values.yaml
Comment (gandro) [inline install/kubernetes/cilium/charts/hubble-ui/values.yaml]: What's the reason for this change? I'm wary of re-introducing the latest tag for hubble-ui. But if we do, please also change the default pull policy to Always.

File: install/kubernetes/cilium/values.yaml
Comment (gandro) [inline install/kubernetes/cilium/values.yaml]: on master, this has to be `latest` I believe. We are only doing non-master tags in release branches.

**Review by gandro — COMMENTED** [review]
Awesome, thanks a lot for doing this! I have a few questions to a few other changes that seem unrelated to the value rename itself.

You will also have to fix the values usage in the CI tests though. See:
https://github.com/cilium/cilium/blob/40320d64b91fc216f516765bb4a62f867aa4a324/test/k8sT/hubble.go#L117

File: install/kubernetes/cilium/charts/hubble-cli/templates/daemonset.yaml
Comment (seanmwinn) [reply install/kubernetes/cilium/charts/hubble-cli/templates/daemonset.yaml]: The flag for `enabled=true` controls whether or not the hubble-cli daemonset manifest is deployed in the cluster. It is a control for the sub-chart. I do see the need to change the test in addition to this.

File: install/kubernetes/cilium/charts/hubble-ui/values.yaml
Comment (seanmwinn) [reply install/kubernetes/cilium/charts/hubble-ui/values.yaml]: this is because `latest` is the only image working in 1.8.0-rc1 with hubble-relay enabled. Note that hubble-ui v0.5.0 also does not work with 1.8.0-rc1 and hubble-relay. We need to cut a new release for this container, tag it, and update this field to something that works.

File: install/kubernetes/cilium/values.yaml
Comment (seanmwinn) [reply install/kubernetes/cilium/values.yaml]: yes, this was not intended - but it does need to reflect the proper version in the release branch which it presently does not

File: install/kubernetes/cilium/values.yaml
Comment (michi-covalent) [inline install/kubernetes/cilium/values.yaml]: i have a question about this, what's the difference between this section and cilium/charts/hubble-cli/values.yaml? if i'm adding a new configuration parameter, do i need to add it to both places?

**Review by michi-covalent — APPROVED** [review]
lgtm, i have a question regarding global vs local values files.

**Review by sayboras — COMMENTED** [review]
You might need to change this as well https://github.com/cilium/cilium/blob/master/install/kubernetes/cilium/requirements.yaml#L10

Just want to express my opinion, I see the benefit of this change, which will help to move one step making sub charts are more self contrained and independent. 

However, there is inter-depedency between sub-charts, and currently we are using _global_ as controlling flag (e.g. hubble-cli chart checks _global.hubble.enabled_). So with change, the subcharts can be using values from other sub-charts directly, I feel this will make it hard to understand and manage.

File: install/kubernetes/cilium/charts/hubble-ui/values.yaml
Comment (michi-covalent) [reply install/kubernetes/cilium/charts/hubble-ui/values.yaml]: @seanmwinn now there is `v0.6.0` tag that is compatible with 1.8.0-rc1/hubble-relay.

File: install/kubernetes/cilium/values.yaml
Comment (seanmwinn) [reply install/kubernetes/cilium/values.yaml]: the sane defaults should be properly located in the sub-charts values files

setting a value at the parent chart level is an override of the default set in the sub-chart

the best practice is to document the settings in the parent chart as commented values

the end-goal is that sane defaults exist throughout the charts and users can maintain a small .yaml file of just their customizations, relying on sane defaults to drive everything else

this will help CI a lot too

Comment (michi-covalent) [reply install/kubernetes/cilium/values.yaml]: makes sense thanks for the explanation!

File: install/kubernetes/cilium/charts/hubble-cli/values.yaml
Comment (aanm) [inline install/kubernetes/cilium/charts/hubble-cli/values.yaml]: This should be kept as IfNotPresent since the tag is not `:latest`

**Review by aanm — CHANGES_REQUESTED** [review]

**Comment by seanmwinn** [comment]
@sayboras Thanks for your feedback. We are trying to focus more on your first point which indicates this will make the charts function in a manner that allows for proper control of sub-charts. The long-term goal is to minimize the use of global values and build sub-charts that are configurable, tunable, and expose the knobs that operators are used to having available such as update strategy, tolerations, annotation, affinity, and others. Today, the parent chart's values file holds little more than configmap values and is not implemented in a manner that I would consider it a production-grade chart.

**Comment by sayboras** [comment]
> Thanks for your feedback. We are trying to focus more on your first point which indicates this will make the charts function in a manner that allows for proper control of sub-charts. The long-term goal is to minimize the use of global values and build sub-charts that are configurable, tunable, and expose the knobs that operators are used to having available such as update strategy, tolerations, annotation, affinity, and others. Today, the parent chart's values file holds little more than configmap values and is not implemented in a manner that meets basic best practices.

@seanmwinn making sense to me :100:, thanks for your explaination :tada:

File: install/kubernetes/cilium/values.yaml
Comment (sayboras) [inline install/kubernetes/cilium/values.yaml]: just curious if we should uncomment PullPolicy (and tag as well), so that _make -C install/kubernetes all_ can update values based on version.

https://github.com/cilium/cilium/blob/master/install/kubernetes/Makefile#L25

File: install/kubernetes/cilium/values.yaml
Comment (sayboras) [inline install/kubernetes/cilium/values.yaml]: what do think about moving these two blocks all the way up, right after _preflight_ subchart values? I think it's easier for user to follow.

File: install/kubernetes/cilium/charts/hubble-relay/values.yaml
Comment (sayboras) [inline install/kubernetes/cilium/charts/hubble-relay/values.yaml]: nits: tag, pullpolicy are still refering to global values. I can send PR later if you are ok.

File: install/kubernetes/cilium/values.yaml
Comment (seanmwinn) [reply install/kubernetes/cilium/values.yaml]: These are resolved as defaults in the sub-chart where you should see them not commented, and thus taking effect. The parent chart is intended for overriding values, and the defaults from the sub-chart are provided as inline comments in the parent chart as a frame of reference.

Comment (sayboras) [reply install/kubernetes/cilium/values.yaml]: thanks, I was thinking that hubble-cli is part of this repo as well (i.e. release cycle is same with cilium), but it is maintained separately.

**Comment by michi-covalent** [comment]
@seanmwinn there is another PR that's related to this: https://github.com/cilium/cilium/pull/11784

if we are including the hubble binary in the cilium docker image, does it mean we can simply get rid of the hubble-cli subchart? let's sync up with @Rolinh .

**Review by michi-covalent — CHANGES_REQUESTED** [review]
i'm blocking this until we clarify if the hubble-cli chart is still needed after https://github.com/cilium/cilium/pull/11784

**Review by michi-covalent — APPROVED** [review]
lgtm, it might conflict with https://github.com/cilium/cilium/pull/11806

File: install/kubernetes/cilium/charts/hubble-relay/values.yaml
Comment (aanm) [inline install/kubernetes/cilium/charts/hubble-relay/values.yaml:8]: @joestringer will this require a change in the Makefile as well? (This looks like a deja vu)

**Review by aanm — CHANGES_REQUESTED** [review]

File: install/kubernetes/cilium/charts/hubble-relay/values.yaml
Comment (joestringer) [reply install/kubernetes/cilium/charts/hubble-relay/values.yaml:8]: yeah. PR #11837 makes the overall changes necessary, we just need to add this extra chart path to the makefile variable. Let's aim to merge that one first then it's an easy change on top.

File: install/kubernetes/cilium/dev-test.yaml
Comment (joestringer) [inline install/kubernetes/cilium/dev-test.yaml]: Is this new file intentionally added? How is it updated?

File: install/kubernetes/cilium/templates/NOTES.txt
Comment (joestringer) [inline install/kubernetes/cilium/templates/NOTES.txt]: Is this change intentional? Most of the PR is about Hubble relay but this change isn't.

**Review by joestringer — CHANGES_REQUESTED** [review]

File: install/kubernetes/cilium/templates/NOTES.txt
Comment (seanmwinn) [reply install/kubernetes/cilium/templates/NOTES.txt]: no this is not intentional, let me clean this up

**Comment by michi-covalent** [comment]
@seanmwinn look like the recent update pulled in a lot of unrelated changes?

File: install/kubernetes/cilium/dev-test.yaml
Comment (seanmwinn) [reply install/kubernetes/cilium/dev-test.yaml]: I made a mistake in a prior commit and rolled this back

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

File: install/kubernetes/cilium/dev-test.yaml
Comment (rolinh) [reply install/kubernetes/cilium/dev-test.yaml]: The file is still there but this conversation is marked as resolved. Based on your comment, I think it was not intentionally added, right?

**Review by rolinh — COMMENTED** [review]
LGTM minus this comment.

File: install/kubernetes/cilium/dev-test.yaml
Comment (aanm) [reply install/kubernetes/cilium/dev-test.yaml]: I missed this file. @seanmwinn please remove it before merging the PR. No need to re-run the CI once the PR is fixed as the tests have passed.

**Review by aanm — CHANGES_REQUESTED** [review]

File: install/kubernetes/cilium/charts/hubble-relay/templates/deployment.yaml
Comment (gandro) [inline install/kubernetes/cilium/charts/hubble-relay/templates/deployment.yaml]: I think you meant to use `Values.image.tag` here.

Comment (seanmwinn) [reply install/kubernetes/cilium/charts/hubble-relay/templates/deployment.yaml]: I did, thanks

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]

**Review by rolinh — APPROVED** [review]

**Pull request merged**
